### PR TITLE
Implement switcher group messaging card

### LINKED ISSUE #16273: Quo2 - Create Switcher - Group Messaging Card
Implement the quo2 component 

https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=911-27084&t=J3iynkJ5hh44GvbG-0

These components already exist in: src/status_im2/contexts/shell/cards/view.cljs

Create the component in isolation to the status mobile best practices.
Add component spec test
Add it's own preview screen (currently shared among others)

### PR DESCRIPTION
**Pull request by codemaster115**
fixes #16273 

### Summary

[comment]: # (Summarise the problem and how the pull request solves it)
Quo2/Group Messaging Card component
https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System-for-Mobile?type=design&node-id=659-24387&mode=design&t=JLx5Xtk08r5cbZ02-4

### Testing notes
You can check in the preview screen

#### Platforms

- Android
- iOS
- macOS
- Linux
- Windows

status: ready

### CODE REVIEW

File: src/quo2/core.cljs
Comment (J-Son89) [inline src/quo2/core.cljs]: why did you move this?  it was alpha sorted before?

File: src/quo2/components/switchers/group_messaging_card/view.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/group_messaging_card/view.cljs]: I think we should avoid reusing styles in many components - either styles are encapsulated in a component and that component is reused or we just duplicate and put these styles in each component 👍

File: src/quo2/components/switchers/one_to_one_chat_card/view.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/one_to_one_chat_card/view.cljs]: This is a TODO or something? 🤔

File: src/quo2/components/switchers/group_messaging_card/view.cljs
Comment (J-Son89) [reply src/quo2/components/switchers/group_messaging_card/view.cljs]: e.g make a component - `card-title` with those styles added and expose that component instead

File: src/quo2/core.cljs
Comment (codemaster115) [reply src/quo2/core.cljs]: Sorry for that. It happened accidentally while rebasing.

File: src/quo2/components/switchers/one_to_one_chat_card/view.cljs
Comment (codemaster115) [reply src/quo2/components/switchers/one_to_one_chat_card/view.cljs]: Yeah, this is TODO for the next PR.

File: src/quo2/components/switchers/base_card/view.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/base_card/view.cljs]: let's just pass `customization-color` along, i.e (style/base-container customization-color)` and handle that in there. I don't think we need this variable

File: src/status_im2/contexts/quo_preview/switcher/base_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/base_card.cljs]: I would say this comment is unnecessary because we know data inside quo previews are always dummy.

File: src/status_im2/contexts/quo_preview/switcher/base_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/base_card.cljs]: You can replace this entire map with a call to `(preview/customization-color-option)`

File: src/status_im2/contexts/quo_preview/switcher/base_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/base_card.cljs]: The style map should be inside the `:style` prop.

File: src/quo2/components/switchers/one_to_one_chat_card/view.cljs
Comment (ilmotta) [reply src/quo2/components/switchers/one_to_one_chat_card/view.cljs]: This namespace has no usages and we don't normally commit blank namespaces like this (I actually don't remember a single instance). This also confuses reviewers (as we can see in this PR) and other devs that might jump into this namespace while the full implementation is not merged. It seems better if we get rid of this ns @codemaster115

File: src/quo2/components/switchers/card_content/view.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/card_content/view.cljs]: The code is pretty clear without this comment IMO.

If you want to signal a problem if no match is found, you can add a call to `(log/warn "Switcher card type not found" {:type type})`, this would be more actionable for a developer. But it seems just returning nil without the comments is quite clear.

File: src/quo2/components/switchers/card_content/view.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/card_content/view.cljs]: Missing `:t` qualification.

File: src/quo2/components/switchers/base_card/view.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/base_card/view.cljs]: This is not a valid form-2 Reagent component. You need to initialize the ratom once. The current code will cause the atom to be initialized on every render of `base-card`.

While doing this, it's a good idea to define the `:ref` function once in the initialization, thus avoiding the unnecessary allocation. Just good practice, not that it matters much.

```clojure
(defn base-card
  []
  (let [card-ref (atom nil)
        set-ref  #(reset! card-ref %)]
    (fn [{:keys [banner on-press on-close customization-color]} & children]
      [rn/touchable-opacity
       {:on-press       on-press
        :ref            set-ref
        :active-opacity 1}
       [rn/view {:style (style/base-container customization-color)}
        (when banner
          [rn/image
           {:source (:source banner)
            :style  {:width 160}}])
        [button/button
         {:size       24
          :type       :grey
          :icon       true
          :on-press   on-close
          :background :photo
          :style      style/close-button}
         :i/close]
        (into [rn/view {:style style/thumb-card}] children)]])))
```

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs]: This zombie code is begging for deletion ;)

**Review by ilmotta — COMMENTED** [review]
I barely tested the component because a designer will run all the checks anyway. Left some comments, but nice work @codemaster115!

**Comment by codemaster115** [comment]
Thanks for your review. @ilmotta @J-Son89 
All comments are resolved now.

File: src/quo2/components/switchers/base_card/view.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/base_card/view.cljs]: @codemaster115, this is not yet a valid form-2 component. The code I shared in this thread should be correct (although I didn't test) https://github.com/status-im/status-mobile/pull/16802#discussion_r1281283245

I suggest the following documentation if this whole form-2 stuff is not clear to you https://github.com/reagent-project/reagent/blob/a14faba55e373000f8f93edfcfce0d1222f7e71a/doc/CreatingReagentComponents.md#form-2--a-function-returning-a-function

File: src/quo2/components/switchers/card_content/view.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/card_content/view.cljs:78]: It's a good practice in status-mobile to always add a default value to `case`, otherwise if, for any reason, `status` is not one of `:unread` or `:mention`, the app will completely crash on the user.

File: src/quo2/components/switchers/utils.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/utils.cljs:6]: Same as another comment I made above, if we don't add a default value, `case` will throw an exception and in production builds this means the app will crash. It's better to have a default value. Better have a `nil` or empty string being returned than a total crash on the user.

File: src/status_im2/contexts/quo_preview/switcher/base_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/base_card.cljs]: Quick one, I didn't spot before. The styles should be inside the `:style` prop. The file src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs has 2 instances of this problem too.

File: src/quo2/components/switchers/base_card/view.cljs
Comment (codemaster115) [reply src/quo2/components/switchers/base_card/view.cljs]: I'm sorry for the duplicate issue again. I will check the suggested documentation to be clear about form-2 component.
Thanks for your patience. @ilmotta

Comment (ilmotta) [reply src/quo2/components/switchers/base_card/view.cljs]: No worries at all @codemaster115 :)

Comment (J-Son89) [reply src/quo2/components/switchers/base_card/view.cljs]: also, don't we wish to start moving toward pressable `api` @ilmotta ?
https://reactnative.dev/docs/pressable

File: ios/Podfile.lock
Comment (J-Son89) [inline ios/Podfile.lock:707]: unless you have updated any dependencies I believe this can be reverted?

File: src/quo2/components/switchers/base_card/style.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/base_card/style.cljs]: this function is technically not "theme" safe and is using the old approach. 
In the old theming approach there was a global ratom which is being used within the theme helper methods (such as this one).
The new theming approach expects that the `theme` prop is passed as a parameter as it leads to safer usage.
This function does not handle theme via prop, so better to use `color/theme-colors` which does. 

However, you'll notice here that these cards are always in `dark` theme so there is no need for theme.

`(colors/custom-color- customization-color 50 40)` will suffice

File: src/quo2/core.cljs
Comment (J-Son89) [inline src/quo2/core.cljs]: `base-card` is internal to the `quo2` module. we should only export 1-1 with Figma from the quo2.core, as such this does not need a preview screen either 👍

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs]: I think we can move away from using the flat list approach in quo2 preview screens. @smohamedjavid wdyt? is it needed here? 🤔

File: src/quo2/components/switchers/card_main_info/style.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/card_main_info/style.cljs]: do we need to set `color` here? these components should only be working in dark theme so we don't need to set items so it works in `light` theme.

File: src/quo2/components/switchers/group_messaging_card/view.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/group_messaging_card/view.cljs]: I think we should call it `color` or `custom-color` or something like that. `-60` seems more like implementation details than describing what it is

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (smohamedjavid) [reply src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs]: It depends on the number of `descriptors`(or options/props) of the component. If it has plenty of `descriptor` like `button` component, this will require a scroll view (we have been using a Flat list for some reason which I don't know) to ensure the content is scrollable in small screen devices too.  

If the component has no descriptor or one/two descriptors, I'm sure it will fit on the screen along with the component below. It all depends on the number of descriptors and the height of the component.

File: src/quo2/components/switchers/group_messaging_card/component_spec.cljs
Comment (ilmotta) [inline src/quo2/components/switchers/group_messaging_card/component_spec.cljs]: Nit: `defonce` is more of a REPL thing to avoid side-effects when evaluating namespaces or the individual def* expressions. In these component spec namespaces on the other hand, they only run successfully via the slow compilation to files, which means, changing the file always cause a complete recompilation and the previous state is lost anyway.

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (ilmotta) [inline src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs:105]: We might as well remove the check for `:code` because this is covered by the default return value.

**Review by ilmotta — APPROVED** [review]

File: src/quo2/components/switchers/group_messaging_card/component_spec.cljs
Comment (codemaster115) [reply src/quo2/components/switchers/group_messaging_card/component_spec.cljs]: So is it still better to use just `def`?

Comment (ilmotta) [reply src/quo2/components/switchers/group_messaging_card/component_spec.cljs]: I would say so @codemaster115. Reserve `defonce` for those cases where reloading in a REPL session can cause unwanted side-effects, otherwise `def` is enough.

Comment (codemaster115) [reply src/quo2/components/switchers/group_messaging_card/component_spec.cljs]: Got it. Thanks, @ilmotta

**Review by Francesca-G — COMMENTED** [review]
Here's the [Figma Frame](https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS/Feedback-for-Mobile?type=design&node-id=3415%3A214488&mode=design&t=fxku3lC2V9UiKVNu-1) with the design review

File: src/quo2/core.cljs
Comment (J-Son89) [inline src/quo2/core.cljs]: Let's keep these Alpha sorted for category too 👍 
This file is huge and easier to keep organised this way

File: src/status_im2/contexts/quo_preview/main.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/main.cljs]: let's keep these sections alpha sorted for ease of using this file 👍

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (J-Son89) [inline src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs:63]: what is this string? is it relevant? probably we can delete it?

**Review by Parveshdhull — APPROVED** [review]
great work :tada:

File: src/quo2/components/switchers/utils.cljs
Comment (J-Son89) [inline src/quo2/components/switchers/utils.cljs:1]: perhaps quo2.components.switchers.labels might be a more apt name? or does this file have more uses?

**Review by J-Son89 — APPROVED** [review]
Looks good @codemaster115! :D

File: src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs
Comment (codemaster115) [reply src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs:63]: The string is just a mock string for the code snippet. @J-Son89

Comment (J-Son89) [reply src/status_im2/contexts/quo_preview/switcher/group_messaging_card.cljs:63]: Oh I see 😅

**Comment by codemaster115** [comment]
The design feedback from @Francesca-G are almost related to the current quo2 component implementation.
Group Avatar, Notification Dot, Counter, and Context Tag are now incompatible.
This PR will be blocked until these components are updated.
@J-Son89 May I create new issues for them?

**Comment by J-Son89** [comment]
@codemaster115 how do you mean they are incompatible? 🤔

Of course let's make new issues if needed, but let's confirm what you mean above first

**Comment by codemaster115** [comment]
This PR is still blocked due to `preview-list` component. 
https://github.com/status-im/status-mobile/pull/17051
Once it's merged, this PR also can be completed.

**Comment by churik** [comment]
required https://github.com/status-im/status-mobile/pull/17051 is merged - please, recheck it after re-base and move to testing when ready, thank you!

**Comment by flexsurfer** [comment]
hey @codemaster115 what's the status of this PR?

**Review by Francesca-G — COMMENTED** [review]
Here's the [design review](https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS/Feedback-for-Mobile?type=design&node-id=3415%3A214488&mode=design&t=PqN01meInlPIXgF0-1) :)

**Review by Francesca-G — COMMENTED** [review]
Here's the [design review](https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS/Feedback-for-Mobile?type=design&node-id=3415%3A214488&mode=design&t=ObUAzrbwsjLrLTDK-1) :)
Please refer to Review 2.

**Comment by codemaster115** [comment]
@Francesca-G Addressed all design reviews except the close button. Because that's related to the implemented quo2 button component. I think it would be better to create a follow-up issue for that.

**Review by Francesca-G — APPROVED** [review]
Looks good now ✨
Adding follow up required due to the "X" button blur, as discussed with @codemaster115 in this [comment](https://www.figma.com/file/Tf5nfkYvpbnNCo4rKLK7lS?node-id=3415:214488&mode=design#569536328)

**Pull request merged**
